### PR TITLE
feat(seer grouping): Store Seer metadata on grouphashes during ingest

### PR DESCRIPTION
**Pull request by lobsterkatie**
This adds the following Seer information to grouphash metadata during ingest:

- When the grouphash was sent to Seer
- Which event's stacktrace was sent (stored as an event id)
- The Seer model version used to analyze the stacktrace
- The matched hash returned by Seer, if any (stored as a reference to that hash's `GroupHash` record)
- The similarity distance returned by Seer, if any

Given that a single event may generate multiple grouphashes, and only one hash value actually gets sent, this also adds a `seer_grouphash_sent` field, which holds a reference to the `GroupHash` corresponding to the sent hash. This not only lets us know which value we should find in the Seer database, it also lets us avoid storing all of the above information redundantly - we can store it only on the `GroupHash` which gets sent, and give the "sibling" `GroupHash`es access to the information, should they need it, via that link.

For example, if an event generates `GroupHash` records `grouphashA` and `grouphashB`, and `grouphashA` is sent:
  - The timestamp, event id, model version, and results will be stored in `grouphashA`'s metadata
  - Both `grouphashA` and `grouphashB` will link to `grouphashA` as the representative which was sent
  - Determining if that event's group has been sent to Seer will be possible by checking for the existence of `metadata.seer_grouphash_sent` on either `grouphashA` or `grouphashB`.
  - The Seer data itself (for example, the send timestamp) will be available via both `grouphashA.seer_grouphash_sent.metadata.seer_date_sent` and `grouphashB.seer_grouphash_sent.metadata.seer_date_sent`.

Because all of this linking makes for a kind of clunky API, at the point at which we actually go to use this information (in grouping info, as a check during future backfills if/when we update our Seer model, etc), we'll probably want to add appropriate properties and/or helper methods to `Group` or `GroupHash`, depending on the application. Until we get to that point, though, it made sense to me to wait rather than guess (potentially incorrectly) what we'll need.

Notes:

- As a result of this change, we're no longer storing Seer results in event data. This is better - before, to see the Seer results you had to find the first event to generate a given hash (which is the first event in a group in cases where Seer doesn't find a match, but is some random event among a group's full event list in cases where Seer does match to an existing group). Now, you can find Seer resuls from any event in a group, via the group's `GroupHash` records. (It did mean I had to pull out some temporary logs I had added to group creation, but it's unclear if they're still necessary. If the issue which prompted them comes up again, I'll add them back in a different way.)

- As agreed offline, I'm updating the `GroupHashMetadata` records which are created elsewhere - rather than waiting to create them until the Seer results are available (or until we decide not to call Seer) - because it's easier to reason about and because the cardinality here is low, given that more than 99% of events match to an existing group and therefore never hit this code. If, after the `GroupHashMetadata` MVP is done, we decide we need to optimize this, we can do so before GA.

- Not done here: Updating the backfill code to store Seer results in grouphash metadata rather than on the group. This should be done before the next time we run a backfill, but given that the current backfill is already half-completed, having them in one place for new groups and a different (single) place for backfilled groups - while not ideal - seemed better than having them in one place for new groups and sometimes the same place but sometimes a different place for backfilled groups.

### CODE REVIEW

File: src/sentry/models/grouphashmetadata.py [hunk 1]
```diff
@@ -4,18 +4,54 @@
...
 
+    # SEER
+
+    # Only one hash representing each group is sent to Seer. For the grouphash actually sent, this
+    # field and the `grouphash` field will be identical. For the grouphashes assigned to the same
+    # group but which aren't sent, this will point to the GroupHash record for the sent hash. Note
+    # that because of merging/unmerging, the sent GroupHash and this metadata's GroupHash (if not
+    # one and the same) aren't guaranteed to forever point to the same group (though they will when
+    # this field is written).
+    seer_grouphash_sent = FlexibleForeignKey(
```
Comment (JoshFerge) [inline src/sentry/models/grouphashmetadata.py:29]: i would prefer not to keep track of this on `GroupHashMetadata`. -- we can simply infer this between the relations between the grouphash and group.

**Review by JoshFerge — COMMENTED** [review]
small note: for when moving out of draft, it would be preferred to isolate the model changes in a singular PR

File: src/sentry/models/grouphashmetadata.py (hunk 1 above)
Comment (JoshFerge) [reply src/sentry/models/grouphashmetadata.py:29]: i see note above on merging/unmerging, i'll think further on this

Comment (lobsterkatie) [reply src/sentry/models/grouphashmetadata.py:29]: Copying here for posterity the highlights of our convo in Slack:

The two situations in which there will be more than one grouphash are some combo of there being a config transition happening and there being separate full-stacktrace and in-app-only hashes. After some investigation/further thinking I realized (confirmed by Armen) that 

a) We thought it was a UI bug that grouping info sometimes shows a green check mark next to both in-app and system, but in fact it’s a logic bug - in-app should take precedence but they both get marked as contributing to grouping. Only contributing hashes make it through to the point where we’re calling Seer, so if we fix the precedence bug, there will only be one contributing hash (and therefore no extra grouphashes to link to the one sent to Seer).

b) The whole point of secondary hashes is to preserve the continuity of an existing group. If there is no existing group, the secondary has no purpose, and similarly therefore shouldn’t make it through to the potential Seer call.

So, TL;DR - I can indeed take that link out, because once we fix the above problems, there will only ever be one hash, meaning nothing to link.

**Comment by lobsterkatie** [comment]
Given that I have to both fix the linking discussed above and split this up into two PRs, I'm going to close this PR in favor of two new ones.

UPDATE: The migration is now handled by https://github.com/getsentry/sentry/pull/78106 and the use of the fields during ingest is handled by https://github.com/getsentry/sentry/pull/78107.
